index.js

```diff
@@ -95,33 +103,33 @@ function registerListener(session, options, cb = () => {}) {
 				totalBytes = 0;
 			}
 
-			if (options.unregisterWhenDone) {
-				session.removeListener('will-download', listener);
-			}
-
-			if (state === 'cancelled') {
-				if (typeof options.onCancel === 'function') {
-					options.onCancel(item);
-				}
-			} else if (state === 'interrupted') {
+			if (state === 'interrupted') {
 				const message = pupa(errorMessage, {filename: item.getFilename()});
 				electron.dialog.showErrorBox(errorTitle, message);
-				cb(new Error(message));
+				reject(new Error(message));
+			} else if (state === 'cancelled') {
+				reject(new Error('The download has been cancelled'));
 			} else if (state === 'completed') {
 				if (process.platform === 'darwin') {
 					app.dock.downloadFinished(filePath);
 				}
 
 				if (options.openFolderWhenDone) {
-					shell.showItemInFolder(path.join(dir, item.getFilename()));
+					shell.showItemInFolder(filePath);
 				}
 
-				cb(null, item);
+				resolve(item);
+			}
+			if (handlerMap.has(url)) {
+				handlerMap.delete(url);
 			}
 		});
 	};
 
-	session.on('will-download', listener);
+	if (!sessionListenerMap.get(session)) {
+		sessionListenerMap.set(session, true);
```

Comment:
This looks like a session leak where we'll never remove this?

Reply:
Don't worry, We just set it one time because of condition `if (!sessionListenerMap.get(session))`

Reply:
I'm not sure what you meant. if we don't have code to remove this. sessionListenerMap would just keep growing infinitely = session leak.

We need to also remove the session when its no longer used?

Reply:
session is dead doesn't mean it'll be removed from the sessionListenerMap, right? So we should listen to window close event and remove session from sessionListenerMap?

Reply:
I updated the code to delete session in sessionListenerMap before windows close event